### PR TITLE
Improve taproot channels support and add tests

### PR DESCRIPTION
**Pull request by t-bast**
This is a PR that targets #3103.

:warning: **Please DO NOT rebase the target branch, and DO NOT partially cherry-pick some of the changes from this PR**: this would require additional efforts to reconcile changes that have been applied from changes that haven't been applied and would waste the large amount of time I spent on reviewing the taproot changes.

I know that this PR contains a lot of commits, but that's on purpose: I spent a lot of time ensuring most of the commits are well-scoped, trivial to review, have a good commit message explaining what is done and the motivation for the change. Please review commit-by-commit and read the commit messages.

By spending such a large amount of time on the taproot changes and improving the tests, I've found a lot of subtle bugs: overall, the "vanilla" scenario works fine with just #3103, but many many edge cases or scenarios that slightly differ from the ideal scenario wasn't working...more details about those bugs can be found in each commit message.

My goal is that we go through the following steps:

- add more commits to this PR until we're satisfied with the result (if necessary)
- at that point, merge it into #3103 and squash the commits into a single "taproot channel support" commit
- then, review and integrate #3118 in `master` (it would be useful to have a review from @pm47)
- and immediately after, integrate the taproot commit on `master` as well: since there is no way to activate the feature yet (unless we modify the code), this should be safe to have on `master`, and will make it much easier to iterate on small fixes/improvements until we have cross-compatibility with `lnd` (it would be a big waste of time to have to handle rebasing with changes on `master`)

I think we should be able to get those done quite soon, which will be a big step forward! After that, I see two important follow-ups that must be done before officially shipping the feature:

- use the final feature bits (we should never ship the "staging" one, that would only create technical debt)
- add support for omitting the `prevtx` field in `interactive-tx` when splicing taproot channels, or when there is a taproot input (such as the swap-in-potentiam input for Phoenix): this will finally get rid of the issue where utxos coming from large transactions (such as coinjoins) cannot be used

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 1]
```diff
@@ -1967,15 +1940,15 @@ class Channel(val nodeParams: NodeParams, val channelKeys: ChannelKeys, val wall
       // Note that if we sent two closing_complete in a row, without waiting for their closing_sig for the first one,
       // this will fail because we only care about our latest closing_complete. This is fine, we should receive their
       // closing_sig for the last closing_complete afterwards.
-      MutualClose.receiveSimpleClosingSig(channelKeys, d.commitments.latest, d.proposedClosingTxs.last, closingSig, remoteClosingNonce_opt, closingCompleteNonces) match {
+      MutualClose.receiveSimpleClosingSig(channelKeys, d.commitments.latest, d.proposedClosingTxs.last, closingSig, localCloserNonces_opt, remoteCloseeNonce_opt) match {
         case Left(f) =>
           log.warning("invalid closing_sig: {}", f.getMessage)
+          remoteCloseeNonce_opt = closingSig.nextCloseeNonce_opt
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:1952]: why update `remoteCloseeNonce_opt` if the closing sig we've received is invalid ?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 2]
```diff
@@ -834,43 +823,44 @@ object Helpers {
...
+              ).flatten
+              closingTxsWithSigs.headOption match {
+                case Some((closingTx, remoteSig, sigToTlv)) =>
+                  val localFundingKey = channelKeys.fundingKey(commitment.fundingTxIndex)
+                  val signedClosingTx_opt = for {
+                    localSig <- closingTx.partialSign(localFundingKey, commitment.remoteFundingPubKey, localNonce, Seq(localNonce.publicNonce, remoteSig.nonce)).toOption
+                    signedTx <- closingTx.aggregateSigs(localFundingKey.publicKey, commitment.remoteFundingPubKey, localSig, remoteSig).toOption
+                  } yield (closingTx.copy(tx = signedTx), localSig.partialSig)
+                  signedClosingTx_opt match {
+                    case Some((signedClosingTx, localSig)) if signedClosingTx.validate(extraUtxos = Map.empty) =>
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:855]: nit: using `Either` instead of `Option` here was a bit clearer

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 3]
```diff
@@ -784,48 +788,33 @@ object Helpers {
...
-              def partialSign(tx: ClosingTx, localNonce: LocalNonce): ChannelSpendSignature.PartialSignatureWithNonce = {
-                val Right(psig) = tx.partialSign(localFundingKey, commitment.remoteFundingPubKey, Map.empty, localNonce, Seq(localNonce.publicNonce, remoteClosingNonce_opt.get))
-                psig
-              }
+            remoteNonce_opt match {
+              case None => return Left(MissingClosingNonce(commitment.channelId))
+              case Some(remoteNonce) =>
+                // If we cannot create our partial signature for one of our closing txs, we just skip it.
+                // It will only happen if our peer sent an invalid nonce, in which case we cannot do anything anyway
+                // apart from eventually force-closing.
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:799]: but it should be left to the caller to decide what do do, here we'll ignore the error and not log anything ?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 4]
```diff
@@ -907,50 +897,40 @@ object Helpers {
...
-                val signedClosingTx = closingTx.copy(tx = signedTx)
-                if (signedClosingTx.validate(extraUtxos = Map.empty)) {
-                  Right(signedClosingTx)
-                } else {
-                  Left(InvalidCloseSignature(commitment.channelId, signedClosingTx.tx.txid))
+                val signedTx = closingTx.aggregateSigs(localFundingKey.publicKey, commitment.remoteFundingPubKey, localSig, remoteSig)
+                Some(closingTx.copy(tx = signedTx))
+              case remoteSig: PartialSignatureWithNonce =>
+                val localNonce = localNonces_opt match {
+                  case Some(localNonces) if closingTx.tx.txOut.size == 2 => localNonces.localAndRemote
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:919]: I slightly prefer the old version with explicit checks on the commitment format instead of building all closing tx (taproot and non-taproot) together.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala [hunk 5]
```diff
@@ -281,70 +272,73 @@ trait ChannelOpenSingleFunded extends SingleFundingHandlers with ErrorHandlers {
...
       Funding.makeFirstCommitTxs(d.channelParams, d.localCommitParams, d.remoteCommitParams, localFundingAmount = 0 sat, remoteFundingAmount = d.fundingAmount, localPushAmount = 0 msat, remotePushAmount = d.pushAmount, d.commitTxFeerate, d.commitmentFormat, fundingTxId, fundingTxOutputIndex, fundingKey, d.remoteFundingPubKey, localCommitmentKeys, remoteCommitmentKeys) match {
-        case Left(ex) => handleLocalError(ex, d, None)
+        case Left(ex) => handleLocalError(ex, d, Some(fc))
         case Right((localSpec, localCommitTx, remoteSpec, remoteCommitTx)) =>
           // check remote signature validity
-          val isRemoveSigValid = fc.sigOrPartialSig match {
-            case psig: ChannelSpendSignature.PartialSignatureWithNonce if d.commitmentFormat.isInstanceOf[SimpleTaprootChannelCommitmentFormat] =>
-              val localNonce = NonceGenerator.verificationNonce(TxId(ByteVector32.Zeroes), fundingKey, 0)
+          val isRemoteSigValid = fc.sigOrPartialSig match {
+            case psig: PartialSignatureWithNonce =>
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala:285]: why remove the commitment format check ? it lets us fail early if the message we've received is invalid

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala [hunk 6]
```diff
@@ -358,18 +352,15 @@ trait ChannelOpenSingleFunded extends SingleFundingHandlers with ErrorHandlers {
...
-        case psig: ChannelSpendSignature.PartialSignatureWithNonce if d.commitmentFormat.isInstanceOf[SimpleTaprootChannelCommitmentFormat] =>
-          val localNonce = NonceGenerator.verificationNonce(TxId(ByteVector32.Zeroes), channelKeys.fundingKey(fundingTxIndex = 0), 0)
-          d.localCommitTx.checkRemotePartialSignature(fundingPubkey, d.remoteFundingPubKey, psig, localNonce.publicNonce)
-        case sig: ChannelSpendSignature.IndividualSignature if !d.commitmentFormat.isInstanceOf[SimpleTaprootChannelCommitmentFormat] =>
-          d.localCommitTx.checkRemoteSig(fundingPubkey, d.remoteFundingPubKey, sig)
-        case _ => false
+      val fundingKey = channelKeys.fundingKey(fundingTxIndex = 0)
+      val isRemoteSigValid = fundingSigned.sigOrPartialSig match {
+        case psig: PartialSignatureWithNonce =>
+          val localNonce = NonceGenerator.verificationNonce(NonceGenerator.dummyFundingTxId, fundingKey, NonceGenerator.dummyRemoteFundingPubKey, 0)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala:358]: same comment about checking the commitment format

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 7]
```diff
@@ -2544,281 +2532,302 @@ class Channel(val nodeParams: NodeParams, val channelKeys: ChannelKeys, val wall
...
-                          case Right(commitSig) => sendQueue = sendQueue :+ commitSig :+ dfu.sharedTx.localSigs
+                        // They haven't received our commit_sig: we retransmit it.
+                        // We're also waiting for signatures from them, and will send our tx_signatures once we receive them.
+                        log.info("re-sending commit_sig for splice attempt with fundingTxIndex={} fundingTxId={}", signingSession.fundingTxIndex, signingSession.fundingTx.txId)
+                        val fundingParams = signingSession.fundingParams
+                        val remoteNonce_opt = channelReestablish.currentCommitNonce_opt
+                        signingSession.remoteCommit.sign(d.commitments.channelParams, signingSession.remoteCommitParams, channelKeys, signingSession.fundingTxIndex, fundingParams.remoteFundingPubKey, signingSession.commitInput(channelKeys), fundingParams.commitmentFormat, remoteNonce_opt) match {
+                          case Left(f) => sendQueue = sendQueue :+ Warning(d.channelId, f.getMessage)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala]: why send a warning and not close the channel ?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 8]
```diff
@@ -207,15 +208,16 @@ case class RemoteCommit(index: Long, spec: CommitmentSpec, txId: TxId, remotePer
...
-      case _: SimpleTaprootChannelCommitmentFormat if remoteNonce_opt.isEmpty =>
-        Left(MissingNonce(channelParams.channelId, commitInput.outPoint.txid))
       case _: SimpleTaprootChannelCommitmentFormat =>
-        val localNonce = NonceGenerator.signingNonce(fundingKey.publicKey)
-        remoteCommitTx.partialSign(fundingKey, remoteFundingPubKey, Map.empty, localNonce, Seq(localNonce.publicNonce, remoteNonce_opt.get)) match {
-          case Left(t) => Left(InvalidNonce(channelParams.channelId, commitInput.outPoint.txid))
-          case Right(psig) => Right(CommitSig(channelParams.channelId, ByteVector64.Zeroes, htlcSigs.toList, TlvStream[CommitSigTlv](CommitSigTlv.PartialSignatureWithNonceTlv(psig))))
+        val localNonce = NonceGenerator.signingNonce(fundingKey.publicKey, remoteFundingPubKey, commitInput.outPoint.txid)
+        remoteNonce_opt match {
+          case Some(remoteNonce) => remoteCommitTx.partialSign(fundingKey, remoteFundingPubKey, localNonce, Seq(localNonce.publicNonce, remoteNonce)) match {
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: create local nonce only if needed:

```suggestion
        remoteNonce_opt match {
          case Some(remoteNonce) =>
            val localNonce = NonceGenerator.signingNonce(fundingKey.publicKey, remoteFundingPubKey, commitInput.outPoint.txid)
            remoteCommitTx.partialSign(fundingKey, remoteFundingPubKey, localNonce, Seq(localNonce.publicNonce, remoteNonce)) match {
```

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 9]
```diff
@@ -660,34 +662,23 @@ case class Commitment(fundingTxIndex: Long,
...
+      case _: SegwitV0CommitmentFormat => remoteCommitTx.sign(fundingKey, remoteFundingPubKey)
       case _: SimpleTaprootChannelCommitmentFormat =>
-        val localNonce = NonceGenerator.signingNonce(fundingKey.publicKey)
-        if (nextRemoteNonce_opt.isEmpty)
-          return Left(MissingNonce(params.channelId, remoteCommitTx.input.outPoint.txid))
-        val Some(remoteNonce) = nextRemoteNonce_opt
-        val psig = remoteCommitTx.partialSign(fundingKey, remoteFundingPubKey, Map.empty, localNonce, Seq(localNonce.publicNonce, remoteNonce)) match {
-          case Left(t) => return Left(InvalidNonce(params.channelId, remoteCommitTx.input.outPoint.txid))
-          case Right(psig) => psig
+        val localNonce = NonceGenerator.signingNonce(fundingKey.publicKey, remoteFundingPubKey, fundingTxId)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: create local nonce after we've checked that there is a remote nonce

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 10]
```diff
@@ -711,36 +702,26 @@ case class Commitment(fundingTxIndex: Long,
   }
 
   /** Return a fully signed commit tx, that can be published as-is. */
-  def fullySignedLocalCommitTx(params: ChannelParams, channelKeys: ChannelKeys): Either[ChannelException, Transaction] = {
+  def fullySignedLocalCommitTx(params: ChannelParams, channelKeys: ChannelKeys): Transaction = {
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala:706]: I'd rather keep returning an `Either` even if it's extremely unlikely that this will fail (data could have been saved with an older version of eclair or our bitcoin libraries for example).

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:1952]: Because if our peer sent us this nonce, that's what they will use for their next signature. If we ignore it, we'll keep getting stuck because their next signature will also be invalid (since we didn't store the correct nonce). Whereas if we store it, the next signature should be correct, because we're kind of "self-healing". It isn't very important though, as we shouldn't run into this anyway, but it's more correct to use the latest nonce.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 2 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:855]: I wasn't a fan of it because the `Left` case was useless (`partialSign` and `aggregateSigs` never return a useful failure). But if you feel strongly about it, don't hesitate to add another commit on top of this branch to go back to an `Either`!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 3 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:799]: No we won't log anything, but it does simplify the code. For an edge case that can only happen if the remote node is buggy, I'd rather keep our code simple, we shouldn't over-complexify our code just to handle cases where our peer is buggy, it's already hairy enough.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:919]: Really? It was duplicating (copy-pasting) the whole logic of how to build each closing signature, which is the hard part...I much prefer that we don't duplicate the closing sig flow, and only pattern match where the two cases differ (when we generate the actual sig or partial sig).

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala (hunk 5 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala:285]: It breaks the completeness of pattern matching if you introduce `if` statements inside the pattern matching clause: you needed a `case _ => false` which is dangerous, because the compiler won't help us whenever we actually need to add a case here. It will fail anyway if the commitment format doesn't match when we call `checkRemotePartialSignature` below, so this has the behavior you're describing?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala (hunk 6 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenSingleFunded.scala:358]: Same as https://github.com/ACINQ/eclair/pull/3136#discussion_r2270651559, and more generally about using complete pattern matching to avoid introducing bugs whenever we introduce more commitment formats.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala (hunk 7 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala]: It seems overkill to force-close immediately: we don't really need to force-close here, so if our peer is buggy it's better to give them a chance to reconnect and fix it instead of wasting a force-close (where we'd be the initiator and thus have to wait for the `to_self_delay` before getting our funds back)? There is a non-zero chance that other implementations don't implement this edge case correctly, and I don't want this to cost us, it should only cost them if they're buggy.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala (hunk 10 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala:706]: Please no, that's actually really important that this cannot fail: once we've verified the remote `commit_sig`, we must always be able to force-close with our local commit. Otherwise we cannot force-close when we actually need to, which is impossible to handle. This is why you had an infinite loop in your previous code, which is really dangerous!

If we have corrupted data, then we're screwed anyway, but any data that we have has been validated when we received it, so I don't see how it could make this code fail?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala (hunk 9 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: Done in https://github.com/ACINQ/eclair/pull/3136/commits/9b4e51d79d664d89806e685aa69c381bd8333806

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala (hunk 8 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: Done in https://github.com/ACINQ/eclair/pull/3136/commits/9b4e51d79d664d89806e685aa69c381bd8333806

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala (hunk 10 above)
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala:706]: What bothers me is that we don't use the same code path to verify incoming signatures and to build the signed commit tx. It should be addressed but it's off-topic here.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 2 above)
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:855]: Will do and add it if it does look better (I agree that it was a bit clumsy).
